design/ErrorHandling.md

```diff
@@ -23,6 +23,141 @@ There are several types of WebGPU calls that get their errors handled differentl
 The extra overhead needs to be low enough that applications can still run while being debugged.
 
+## *Fatal Errors*: Lost/Recovered Events
+
+<!-- calling this revision 5 -->
+
+```webidl
+partial interface GPUAdapter {
+    Promise<GPUDevice> requestDevice(GPUDeviceDescriptor descriptor,
+                                     GPUDeviceLostCallback onDeviceLost);
+};
+
+callback GPUDeviceLostCallback = void (GPUDeviceLostInfo info);
+
+interface GPUDeviceLostInfo {
+    readonly attribute GPUDevice device;
+    readonly attribute DOMString reason;
+};
+```
+
+`GPUAdapter.requestDevice` requests a device from the adapter.
+It returns a Promise which resolves when a device is ready.
+The Promise may not resolve for a long time - it resolves when the browser is ready for the application to bring up (or restore) its content.
+If the adapter is lost before the Promise resolves, the Promise rejects.
+
+`requestDevice` takes a required `onDeviceLost` callback.
+When it is called, the `GPUDevice` cannot be used anymore.
+The device and all objects created from the device have become invalid.
+All further operations on the device and its objects are errors.
+The `"gpulogentry"` event will no longer fire. (This makes all further operations no-ops.)
+(TODO: this could probably be an `ondevicelost` event instead.
+However, this (1) makes it required and (2) makes it clear that no error can be missed between creation and event listener registration.)
+
+`onDeviceLost` may be called if something goes fatally wrong on the device (e.g. unexpected out-of-memory, crash, or native device loss).
+When the application receives this callback, it may immediately request a new device.
+
+### Example Code
+
+```js
+class MyRenderer {
+  constructor() {
+    this.rendering = false;
+    this.adapter = null;
+    this.device = null;
+  }
+  async begin() {
+    try {
+      await this.initWebGPU();
+    } catch (e) {
+      console.error(e);
+      this.initFallback();
+    }
+  }
+  async initWebGPU() {
+    await this.ensureDevice();
+    // ... Upload resources, etc.
+  }
+  initFallback() { /* ... */ }
+  async ensureDevice() {
+    // Stop rendering. (If there was already a device, WebGPU calls made before
+    // the app notices the device is lost are okay - they are no-ops.)
+    this.rendering = false;
+    if (!this.adapter) {
+      // If no adapter, get one.
+      // (If requestAdapter rejects, no matching adapter is available. Exit to fallback.)
+      this.adapter = await gpu.requestAdapter({ /* options */ });
+    }
+    try {
+      this.device = await this.adapter.requestDevice({ /* options */ }, (info) => {
+        // Device was lost.
+        console.error("device lost", info);
+        // Try to get a device again.
+        this.ensureDevice();
+      });
+      this.rendering = true;
+    } catch (e) {
+      // Request failed (likely due to adapter loss).
```

Comment:
I actually considered and specifically avoided this additional signal. I don't think there is any case in which the application would need to know about an adapter loss, aside from knowing that the device(s) were lost. They can try adapter.createDevice to discover whether the adapter will no longer vend any new devices.

Also, since one device on an adapter can be lost without other devices on the same adapter being lost, the signal needed to be on the device.